### PR TITLE
feat(ct): before mount hook wrapper

### LINKED ISSUE #14345: [Feature] Allow custom mount fixture for component testing
It would be great to be able to use pattern well adopted in React Testing Library tests - a custom render function which contains wrappers required by each component (ie. providers). Let's take a look at the example:

```tsx
// Button.test.tsx
import { test as base, expect } from '@playwright/experimental-ct-react';
import { CustomWrapper } from 'src/CustomWrapper';
import { Button } from './Button';

const test = base.extend({
  mount: async ({ mount: baseMount }, use) => {
    const mount: typeof baseMount = (component) => baseMount(<CustomWrapper>{component}</CustomWrapper>);

    await use(mount);
  }
})

test('works', async ({ mount }) => {
  const component = await mount(<Button>a button</Button>);

  await expect(await component.screenshot()).toMatchSnapshot();
})
```

The above example works, because everything is implemented within the test module. This pattern, however, is not working when extracting extended `test` object to a separate module like so:

```tsx
// componentTest.tsx
import { test as base } from '@playwright/experimental-ct-react';
import { CustomWrapper } from 'src/CustomWrapper';

const test = base.extend({
  mount: async ({ mount: baseMount }, use) => {
    const mount: typeof baseMount = (component) => baseMount(<CustomWrapper>{component}</CustomWrapper>);

    await use(mount);
  }
})

export * from '@playwright/experimental-ct-react';
export { test };
```

```tsx
// Button.test.tsx
import { test, expect } from 'src/componentTest'; // <----- this imports resolve to strings
import { Button } from './Button';

test('works', async ({ mount }) => {
  const component = await mount(<Button>a button</Button>);

  await expect(await component.screenshot()).toMatchSnapshot();
})
```

I assume this is due to custom `tsxTransform.ts` that keeps track of every component/jsx module that is imported within test module.

SIDE NOTE: Even though I can't really figure out if it's possible to easily workaround this, in order to actually be able to test I found a hacky way to achieve this result by monkey-patching `react-dom` like so:

```tsx
// react-dom.patched.tsx
import _ReactDOM from '../node_modules/react-dom';
import { CustomWrapper } from './CustomWrapper';

const ReactDOM = Object.create(_ReactDOM);

ReactDOM.render = (children, el) => _ReactDOM.render(<CustomWrapper>{children}</CustomWrapper>, el);

export default ReactDOM;
```

and then I alias `react-dom` in Vite configuration that is passed to Playwright.

What do you guys think about it?

### PR DESCRIPTION
**Pull request by sand4rt**
- [x] Solid before mount
- [x] React before mount
- [x] React vite before mount
- [x] Vue vite before mount alignment
- [x] Vue cli before mount alignment
- [x] Vue2 before mount alignment
~~Svelte before mount~~
~~Svelte vite before mount~~

Not sure if a similar pattern can be devised for Svelte. It seems that this is not possible.

@pavelfeldman @mlmmn this closes https://github.com/microsoft/playwright/issues/14345 right?

### CODE REVIEW

File: tests/components/ct-solid/playwright/index.tsx [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
+import { Router } from "@solidjs/router";
 
 export type HooksConfig = {
   route: string;
 }
 
-beforeMount<HooksConfig>(async ({ hooksConfig }) => {
+beforeMount<HooksConfig>(async ({ hooksConfig, Component }) => {
   console.log(`Before mount: ${JSON.stringify(hooksConfig)}`);
+  return <Router><Component /></Router>;
```
Comment (sand4rt) [inline tests/components/ct-solid/playwright/index.tsx:11]: @pavelfeldman Thinking out loud: Maybe it's useful that the plugins/wrappers for the JSX frameworks can be registered globally in the beforeEach hook, as this is also possible with Vue. This ensures that the tests themselves are less framework specific.

File: tests/components/ct-solid/playwright/index.tsx [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
+import { Router } from "@solidjs/router";
 
 export type HooksConfig = {
   route: string;
 }
 
-beforeMount<HooksConfig>(async ({ hooksConfig }) => {
+beforeMount<HooksConfig>(async ({ hooksConfig, App }) => {
   console.log(`Before mount: ${JSON.stringify(hooksConfig)}`);
+  return <Router><App /></Router>;
```
Comment (sand4rt) [inline tests/components/ct-solid/playwright/index.tsx:11]: @pavelfeldman Thinking out loud: Maybe it's useful that the plugins/wrappers for the JSX frameworks can be registered globally in the beforeEach hook, as this is also possible with Vue: `app.use(router)`. This ensures less code duplication and the tests themselves are less framework specific.

Comment (pavelfeldman) [reply tests/components/ct-solid/playwright/index.tsx:11]: Exposing wrapping facility certainly works for me. We can even make it explicit and introduce a third hook for it, but return value from the beforeMount is a good-enough start.

Comment (sand4rt) [reply tests/components/ct-solid/playwright/index.tsx]: Cool! Will create a follow up PR for the others

Comment (sand4rt) [reply tests/components/ct-solid/playwright/index.tsx]: Cool, i will develop it a bit further. I think a conscious choice has to be made how to deal with

Explored other possibilities:

## 1. `beforeMount` method 

### Solid
```tsx
import { beforeMount } from '@playwright/experimental-ct-solid/hooks';

beforeMount(async ({ App }) => {
  if (hooksConfig?.routing) return <Router><App /></Router>;
});
```

```tsx
test('works', async ({ mount }) => {
  await mount(<App />, { hooksConfig: { routing: true } });
});
```

### Vue

```tsx
import { beforeMount } from '@playwright/experimental-ct-solid/hooks';

beforeMount(async ({ app }) => {
  if (hooksConfig?.routing) app.use(router);
});
```

```tsx
test('works', async ({ mount }) => {
  await mount(<App />, { hooksConfig: { routing: true } });
});
```

---

## 2. `base.extend` method

### Solid
```tsx
import { test: base } from '@playwright/experimental-ct-solid';

const test = base.extend({
  mount({ mount }, use) => use((App) => mount(<Router><App /></Router>));
});
```

```tsx
test('works', async ({ mount }) => {
  await mount(<App />});
});
```


---


## 3. `beforeEach` method
### Solid
```tsx
import { test } from '@playwright/experimental-ct-solid';

test.beforeEach(async ({ App }) => <Router><App /></Router>);
```

```tsx
test('works', async ({ mount }) => {
  await mount(<Component />});
});
```

---


## 4. `factory` method

### Solid

```tsx
import { mount } from '@playwright/experimental-ct-solid';

function mountWithRouter(overrides = {}) {
  return mount(<Router><Component /></Router>)
}
```

```tsx
test('works', async () => {
  await mountWithRouter();
});

```

Comment (sand4rt) [reply tests/components/ct-solid/playwright/index.tsx]: Cool, will develop it a bit further! Explored some other possibilities as well. Takes a while to explain but _(without looking at technical limitations)_ i think the `beforeMount` hook is the best option of all.

## 1. `beforeMount` method 

```tsx
import { beforeMount } from '@playwright/experimental-ct-solid/hooks';

beforeMount(async ({ App }) => {
  if (hooksConfig?.routing) return <Router><App /></Router>;
});
```

```tsx
test('works', async ({ mount }) => {
  await mount(<App />, { hooksConfig: { routing: true } });
});
```

## 2. `base.extend` method

```tsx
import { test: base } from '@playwright/experimental-ct-solid';

const test = base.extend({
  mount({ mount }, use) => use((App) => mount(<Router><App /></Router>));
});
```

## 3. `beforeEach` method

```tsx
import { test } from '@playwright/experimental-ct-solid';

test.beforeEach(async ({ mount }) => mount(<Router><App /></Router>));
```

```tsx
test('works', async ({ mount }) => {
  await mount(<Component />});
});
```

## 4. `factory` method

```tsx
import { mount } from '@playwright/experimental-ct-solid';

function mountWithRouter(overrides = {}) {
  return mount(<Router><Component /></Router>)
}
```

```tsx
test('works', async () => {
  await mountWithRouter();
});
```

Comment (pavelfeldman) [reply tests/components/ct-solid/playwright/index.tsx]: I'm with you, beforeMount is the most straightforward and powerful given that it runs in the browser.

**Comment by pastelsky** [comment]
Just want to say that having this would be super helpful for storybook-style global decoration. Looking forward to this making its way to the core.

**Comment by pastelsky** [comment]
Looks like all pre-requisites are done for this one.

**Comment by sand4rt** [comment]
> Looks like all pre-requisites are done for this one.

I'm on it, hope to have the PR ready for review this week

File: packages/playwright-ct-react/registerSource.mjs [hunk 3]
```diff
@@ -69,10 +70,13 @@ function render(component) {
 }
 
 window.playwrightMount = async (component, rootElement, hooksConfig) => {
-  for (const hook of /** @type {any} */(window).__pw_hooks_before_mount || [])
-    await hook({ hooksConfig });
+  let App = () => render(component);
+  for (const hook of /** @type {any} */(window).__pw_hooks_before_mount || []) {
+    const wrapper = await hook({ App, hooksConfig });
+    if (wrapper) App = () => wrapper;
```
Comment (pavelfeldman) [inline packages/playwright-ct-react/registerSource.mjs:76]: nit: style, if expression goes to the next line.

**Review by pavelfeldman — APPROVED** [review]

File: packages/playwright-ct-react/registerSource.mjs (hunk 3 above)
Comment (sand4rt) [reply packages/playwright-ct-react/registerSource.mjs]: fixed

**Pull request merged**